File: tools/audiotap/Tests/StreamingMasterResamplerTests.swift
@testable import AudioTapLib
import XCTest

@available(macOS 14.2, *)
final class StreamingMasterResamplerTests: XCTestCase {
    func test44100To48000TracksRationalCountAcrossTenThousandCallbacks() throws {
        let converter = try XCTUnwrap(
            StreamingMasterResampler(
                inputRate: 44_100,
                inputChannels: 1,
                outputRate: 48_000,
                outputChannels: 1,
            ),
        )
        var outputFrames: Int64 = 0
        var firstPhase: Double?
        var secondPhase: Double?

        for callback in 0 ..< 10_000 {
            outputFrames += Int64(
                converter.process([Float](repeating: 0.25, count: 256)).count,
            )
            if callback == 0 {
                firstPhase = converter.fractionalPhase
            } else if callback == 1 {
                secondPhase = converter.fractionalPhase
            }
        }
        outputFrames += Int64(converter.finish().count)

        let expected = roundedRatio(
            inputFrames: 10_000 * 256,
            inputRate: 44_100,
            outputRate: 48_000,
        )
        let firstPhase = try XCTUnwrap(firstPhase)
        let secondPhase = try XCTUnwrap(secondPhase)
        XCTAssertEqual(outputFrames, converter.totalOutputFrames)
        XCTAssertLessThanOrEqual(abs(outputFrames - expected), 1)
        XCTAssertNotEqual(firstPhase, 0, accuracy: 0.000000000001)
        XCTAssertNotEqual(firstPhase, secondPhase, accuracy: 0.000000000001)
    }

    func test48000To44100TracksRationalCountAcrossTenThousandCallbacks() throws {
        let converter = try XCTUnwrap(
            StreamingMasterResampler(
                inputRate: 48_000,
                inputChannels: 1,
                outputRate: 44_100,
                outputChannels: 1,
            ),
        )
        var outputFrames: Int64 = 0

        for _ in 0 ..< 10_000 {
            outputFrames += Int64(
                converter.process([Float](repeating: -0.25, count: 127)).count,
            )
        }
        outputFrames += Int64(converter.finish().count)

        let expected = roundedRatio(
            inputFrames: 10_000 * 127,
            inputRate: 48_000,
            outputRate: 44_100,
        )
        XCTAssertEqual(outputFrames, converter.totalOutputFrames)
        XCTAssertLessThanOrEqual(abs(outputFrames - expected), 1)
    }

    func testAlternatingCallbackSizesKeepsCumulativeCount() throws {
        let converter = try XCTUnwrap(
            StreamingMasterResampler(
                inputRate: 44_100,
                inputChannels: 1,
                outputRate: 48_000,
                outputChannels: 1,
            ),
        )
        let callbackSizes = [127, 256, 511]
        var inputFrames: Int64 = 0
        var outputFrames: Int64 = 0

        for callback in 0 ..< 3_000 {
            let size = callbackSizes[callback % callbackSizes.count]
            inputFrames += Int64(size)
            outputFrames += Int64(
                converter.process([Float](repeating: 0.1, count: size)).count,
            )
        }
        outputFrames += Int64(converter.finish().count)

        let expected = roundedRatio(
            inputFrames: inputFrames,
            inputRate: 44_100,
            outputRate: 48_000,
        )
        XCTAssertEqual(converter.totalInputFrames, inputFrames)
        XCTAssertEqual(outputFrames, converter.totalOutputFrames)
        XCTAssertLessThanOrEqual(abs(outputFrames - expected), 1)
    }

    func testRampSplitAcrossCallbacksMatchesOneBlockAndStaysContinuous() throws {
        let inputRate = 44_100
        let outputRate = 48_000
        let ramp = (0 ..< 4_096).map { Float($0) / 4_096 }
        let oneBlock = try output(
            ramp,
            inputRate: inputRate,
            outputRate: outputRate,
            callbackSizes: [ramp.count],
        )
        let split = try output(
            ramp,
            inputRate: inputRate,
            outputRate: outputRate,
            callbackSizes: [127, 256, 511, 89],
        )

        XCTAssertEqual(split.count, oneBlock.count)
        let maximumSplitError = zip(split, oneBlock)
            .map { abs($0 - $1) }
            .max() ?? 0
        XCTAssertLessThan(maximumSplitError, 0.00002)

        let maximumBoundaryDelta = zip(split.dropFirst(), split)
            .map { abs($0 - $1) }
            .max() ?? 0
        XCTAssertLessThan(maximumBoundaryDelta, 0.002)
    }

    private func output(
        _ samples: [Float],
        inputRate: Int,
        outputRate: Int,
        callbackSizes: [Int],
    ) throws -> [Float] {
        let converter = try XCTUnwrap(
            StreamingMasterResampler(
                inputRate: inputRate,
                inputChannels: 1,
                outputRate: outputRate,
                outputChannels: 1,
            ),
        )
        var result: [Float] = []
        var offset = 0
        var callback = 0
        while offset < samples.count {
            let requested = callbackSizes[callback % callbackSizes.count]
            let count = min(requested, samples.count - offset)
            result += converter.process(Array(samples[offset ..< (offset + count)]))
            offset += count
            callback += 1
        }
        result += converter.finish()
        XCTAssertEqual(Int64(result.count), converter.totalOutputFrames)
        return result
    }

    private func roundedRatio(
        inputFrames: Int64,
        inputRate: Int,
        outputRate: Int,
    ) -> Int64 {
        let numerator = inputFrames * Int64(outputRate)
        let denominator = Int64(inputRate)
        return numerator / denominator
            + ((numerator % denominator) * 2 >= denominator ? 1 : 0)
    }
}

File: app/ClassScribe.Windows/tests/ClassScribe.Core.Tests/StreamingMasterResamplerTests.cs
using ClassScribe.Core;

namespace ClassScribe.Core.Tests;

[TestClass]
public sealed class StreamingMasterResamplerTests
{
    [TestMethod]
    public void resampler44100To48000TracksRationalCountAcrossTenThousandCallbacks()
    {
        var converter = new StreamingMasterResampler(44_100, 1, 48_000, 1);
        long outputFrames = 0;
        double firstPhase = 0;
        double secondPhase = 0;

        for (var callback = 0; callback < 10_000; callback++)
        {
            outputFrames += converter.Process(new float[256]).Length;
            if (callback == 0)
            {
                firstPhase = converter.FractionalPhase;
            }
            else if (callback == 1)
            {
                secondPhase = converter.FractionalPhase;
            }
        }

        outputFrames += converter.Finish().Length;
        var expected = RoundedRatio(10_000L * 256, 44_100, 48_000);

        Assert.AreEqual(outputFrames, converter.TotalOutputFrames);
        Assert.IsTrue(Math.Abs(outputFrames - expected) <= 1);
        Assert.AreNotEqual(0, firstPhase, 0.000000000001);
        Assert.AreNotEqual(firstPhase, secondPhase, 0.000000000001);
    }

    [TestMethod]
    public void resampler48000To44100TracksRationalCountAcrossTenThousandCallbacks()
    {
        var converter = new StreamingMasterResampler(48_000, 1, 44_100, 1);
        long outputFrames = 0;

        for (var callback = 0; callback < 10_000; callback++)
        {
            outputFrames += converter.Process(new float[127]).Length;
        }

        outputFrames += converter.Finish().Length;
        var expected = RoundedRatio(10_000L * 127, 48_000, 44_100);

        Assert.AreEqual(outputFrames, converter.TotalOutputFrames);
        Assert.IsTrue(Math.Abs(outputFrames - expected) <= 1);
    }

    [TestMethod]
    public void alternatingCallbackSizesKeepCumulativeCount()
    {
        var converter = new StreamingMasterResampler(44_100, 1, 48_000, 1);
        var sizes = new[] { 127, 256, 511 };
        long inputFrames = 0;
        long outputFrames = 0;

        for (var callback = 0; callback < 3_000; callback++)
        {
            var size = sizes[callback % sizes.Length];
            inputFrames += size;
            outputFrames += converter.Process(new float[size]).Length;
        }

        outputFrames += converter.Finish().Length;
        var expected = RoundedRatio(inputFrames, 44_100, 48_000);

        Assert.AreEqual(inputFrames, converter.TotalInputFrames);
        Assert.AreEqual(outputFrames, converter.TotalOutputFrames);
        Assert.IsTrue(Math.Abs(outputFrames - expected) <= 1);
    }

    [TestMethod]
    public void rampSplitAcrossCallbacksMatchesOneBlockAndStaysContinuous()
    {
        const int inputRate = 44_100;
        const int outputRate = 48_000;
        var ramp = Enumerable.Range(0, 4_096)
            .Select(index => index / 4_096f)
            .ToArray();

        var oneBlock = Collect(ramp, inputRate, outputRate, [ramp.Length]);
        var split = Collect(ramp, inputRate, outputRate, [127, 256, 511, 89]);

        Assert.AreEqual(oneBlock.Length, split.Length);
        var maximumSplitError = oneBlock
            .Zip(split)
            .Select(pair => Math.Abs(pair.First - pair.Second))
            .DefaultIfEmpty()
            .Max();
        Assert.IsTrue(maximumSplitError < 0.00002f);

        var maximumBoundaryDelta = split
            .Zip(split.Skip(1))
            .Select(pair => Math.Abs(pair.Second - pair.First))
            .DefaultIfEmpty()
            .Max();
        Assert.IsTrue(maximumBoundaryDelta < 0.002f);
    }

    [TestMethod]
    public void captureHandoffTimelineUsesActualStreamingFrameCount()
    {
        var root = Path.Combine(
            Path.GetTempPath(),
            $"classscribe-streaming-timeline-{Guid.NewGuid():N}");
        Directory.CreateDirectory(root);
        try
        {
            var writer = new MasterAudioWriter(Path.Combine(root, "audio-manifest.json"));
            var attempt = SessionAttemptID.Create(Guid.NewGuid(), 1);
            var firstGeneration = new CaptureSourceGeneration(attempt, 1);
            var secondGeneration = new CaptureSourceGeneration(attempt, 2);
            var firstFormat = AudioPcmFormat.Create(
                48_000,
                2,
                AudioSampleEncoding.Float32LE);
            var secondFormat = AudioPcmFormat.Create(
                44_100,
                1,
                AudioSampleEncoding.PcmS16LE);
            var first = writer.PreparePacket(
                PcmAudioConverter.EncodeFloat32LE([0.1f, -0.1f, 0.2f, -0.2f]),
                firstFormat,
                firstGeneration);
            var timeline = new CaptureHandoffTimeline();
            timeline.SetFormat(first.Format);
            timeline.BeginGeneration(firstGeneration);
            var firstPlan = timeline.PreparePacket(
                firstGeneration,
                first.Bytes.Length,
                arrivalTimestamp: () => 0);
            timeline.CommitPacket(firstPlan);
            writer.CommitFrames(first.FrameCount);

            var oldTail = writer.FlushPendingPacket();
            writer.CommitFrames(oldTail.FrameCount);
            timeline.AdvanceDurableFrames(oldTail.FrameCount);
            timeline.MarkHandoff(secondGeneration);

            var second = writer.PreparePacket(
                PcmAudioConverter.EncodePcmS16LE([0.3f, -0.3f, 0.4f, -0.4f]),
                secondFormat,
                secondGeneration);
            var arrival = timeline.Snapshot.LastPacketEndTimestamp ?? 0;
            var secondPlan = timeline.PreparePacket(
                secondGeneration,
                second.Bytes.Length,
                arrivalTimestamp: () => arrival);

            Assert.IsFalse(second.IsEmpty);
            Assert.AreEqual(second.FrameCount * second.Format.BytesPerFrame, second.Bytes.Length);
            Assert.AreEqual(second.Bytes.Length, secondPlan.ByteCount);
        }
        finally
        {
            if (Directory.Exists(root))
            {
                Directory.Delete(root, recursive: true);
            }
        }
    }

    private static float[] Collect(
        float[] samples,
        int inputRate,
        int outputRate,
        int[] callbackSizes)
    {
        var converter = new StreamingMasterResampler(inputRate, 1, outputRate, 1);
        var result = new List<float>();
        var offset = 0;
        var callback = 0;
        while (offset < samples.Length)
        {
            var requested = callbackSizes[callback % callbackSizes.Length];
            var count = Math.Min(requested, samples.Length - offset);
            result.AddRange(converter.Process(samples.AsSpan(offset, count)));
            offset += count;
            callback++;
        }

        result.AddRange(converter.Finish());
        Assert.AreEqual((long)result.Count, converter.TotalOutputFrames);
        return result.ToArray();
    }

    private static long RoundedRatio(long inputFrames, int inputRate, int outputRate)
    {
        var numerator = checked(inputFrames * outputRate);
        var quotient = numerator / inputRate;
        var remainder = numerator % inputRate;
        return quotient + (remainder * 2 >= inputRate ? 1 : 0);
    }
}
